### PR TITLE
Makes QueryGraph's edge-explorer cache work for all EdgeFilters, fixes #1623.

### LINKED ISSUE #1623: QueryGraph edge-explorer cache does not work for edge-based CH.
We can do 
```java
queryGraph.setUseEdgeExplorerCache(true);
```
and 
```java
queryGraph.createEdgeExplorer(edgeFilter)
```
will return the same edge-explorer instance for all subsequent calls of `DefaultEdgeFilter.inEdges`, `DefaultEdgeFilter.outEdges`, `DefaultEdgeFilter.allEdges` and `EdgeFilter.ALL_EDGES` (for each of these filters there will be a separate edge-explorer instance).

The problem is that in some cases we explicitly need to create two different instances for the same filter. This is not needed very often, but for example here: 

https://github.com/graphhopper/graphhopper/blob/d24173a188f4574b6eb9e7c6ff641eaec9029208/core/src/main/java/com/graphhopper/routing/AbstractBidirectionEdgeCHNoSOD.java#L46-L48

there is a nested loop over edge iterators where the 'inner' explorer must be different to the 'outer' one.

Because of this using the edge-explorer cache with edge-based CH fails at the moment.

I started adding a test for the query graph behavior here: https://github.com/graphhopper/graphhopper/commit/58e7ac1777c2382d5a42dd7c9153c2de65e31a28, but so far it does not show much, obviously if we use the same instance of the edge explorer within an inner loop the outer loop will be messed up.

### PR DESCRIPTION
**Pull request by easbar**
1) `QueryGraph` caches edge-explorers based on the identity of the `EdgeFilter`s passed into `createEdgeExplorer`
2) fixes edge-based CH in combination with edge-explorer cache (by setting explicit filter ids)
3) deprecates creators of `DefaultEdgeFilter` using `FlagEncoder` (and adds new ones taking only access encoded value instead)
4) replaces `PrepEdgeFilter` with `DefaultEdgeFilter` in `PrepareRoutingSubnetworks` (it did not really do anything different than `DefaultEdgeFilter#allEdges`).

Most changed files are due to 3), and not sure about this. 
- good: `DefaultEdgeFilter` is decoupled from `FlagEncoder`
- bad: it changes a frequently used API and its a bit less comfortable because in most cases we have to pass `flagEncoder.getAccessEnc()` instead of simply `flagEncoder`. Also 3) is a bit unrelated to the actual issue #1623. 

WDYT ? I can also revert this.

### CODE REVIEW

**Comment by karussell** [comment]
Thanks, this looks good.

Regarding the deprecation: this would be nice if the actual code amount reduces somehow. We would need to get rid of the FlagEncoder usage of the calling side first i.e. instead of encodingManager.getFlagEncoder("xy").getAccessEnc() we would do encodingManager.getEncodedValue("car.access") or something. Probably we should indeed move this into a separate issue as this is more related to reducing the usage of FlagEncoder. I would still provide the new method for better migration, just remove the deprecation?

Furthermore I would remove the acceptsForward&Backwards methods as I cannot find a usage, or do you think this would be too ugly?

**Comment by easbar** [comment]
> to get rid of the FlagEncoder usage of the calling side first 

I checked and in many cases this is what changed: 

```java
- outEdgeExplorer = graph.createEdgeExplorer(DefaultEdgeFilter.outEdges(flagEncoder));
+ outEdgeExplorer = graph.createEdgeExplorer(DefaultEdgeFilter.outEdges(flagEncoder.getAccessEnc()));
```
Typically the `flagEncoder` field of the classes where this happens is not used for anything else and the `flagEncoder` field could be replaced with a boolean encoded value there already. But I agree this is not the right place to do such refactoring.

>  Probably we should indeed move this into a separate issue as this is more related to reducing the usage of FlagEncoder. I would still provide the new method for better migration, just remove the deprecation?

Ok yes any further changes should go in a separate issue, but should we keep the changes (like the ones above) or do you want to revert this ? Either way is fine for me. If we keep the changes why not keep the deprecation ? The method taking a flag encoder is no longer used, but I thought it would be easier to adopt downstream when the method is kept and only deprecated. If we change back to `.in/outEdges(encoder)` everywhere sure we should also remove the `@Deprecated` annotations.

> Furthermore I would remove the acceptsForward&Backwards methods as I cannot find a usage, or do you think this would be too ugly?

Ok.

**Comment by karussell** [comment]
> Typically the flagEncoder field of the classes where this happens is not used for anything else 

Yes, exactly. I would either reduce the FlagEncoder usage also on the calling side (new issue) or keep using the old and shorter variant. Still it would be also no problem if you keep it to save time and schedule the changes for later :)

> If we keep the changes why not keep the deprecation?

Good question - convinced ;)

**Comment by easbar** [comment]
Ok I reverted to the old shorter variant and removed the deprecation.

**Comment by easbar** [comment]
The build failed with
```
Minutely API limit heavily violated: 300.6 >= 200.0. Find the details at the bottom of the page https://www.graphhopper.com/pricing/
```

File: core/src/main/java/com/graphhopper/routing/QueryGraph.java [hunk 1]
```diff
@@ -57,8 +57,7 @@
     private final int mainEdges;
     private final QueryGraph baseGraph;
     private final GraphExtension wrappedExtension;
-    // TODO when spreading it on different threads we need multiple independent explorers
```
[the remark this answers is not in the document]
Comment (karussell) [reply core/src/main/java/com/graphhopper/routing/QueryGraph.java]: Yes, this can be removed.

**Comment by karussell** [comment]
> The build failed with

Oh, likely too many tests in parallel (maybe also some misusage). Have increased this limit for now and triggered the build again.

**Comment by easbar** [comment]
I moved the (now reverted) flagEncoder vs encoded value refactoring code changes to here: https://github.com/graphhopper/graphhopper/issues/1625

**Pull request merged**